src/SymbolTable.ts

```diff
@@ -0,0 +1,97 @@
+import type { Identifier } from './lexer/Token';
+import type { BscType } from './types/BscType';
+import { DynamicType } from './types/DynamicType';
+import { UninitializedType } from './types/UninitializedType';
+
+export interface BscSymbol {
+    name: Identifier;
+    type: BscType;
+}
+
+/**
+ * Stores the types associated with variables and functions in the Brighterscript code
+ * Can be part of a hierarchy, so lookups can reference parent scopes
+ */
+export class SymbolTable {
+
+    private symbols = new Map<string, BscSymbol[]>();
+
+    constructor(
+        private parent?: SymbolTable | undefined
+    ) { }
+
+    /**
+     * Sets the parent table for lookups
+     *
+     * @param {SymbolTable} [parent]
+     */
+    setParent(parent?: SymbolTable) {
+        this.parent = parent;
+    }
+
+    /**
+     * Gets the name/type pair for a given named variable or function name
+     * If the identifier is not in this table, it will check the parent
+     *
+     * @param {string)} name the name to lookup
+     * @returns {BscSymbol}
+     */
+    getSymbol(name: string): BscSymbol[] {
+        const key = name.toLowerCase();
+        return this.symbols.get(key) ?? this.parent?.getSymbol(key);
+    }
+
+    /**
+     * Adds a new symbol to the table
+     * @param {Identifier} name
```

Comment:
You don't need to specify the data type (i.e. `{Identifier}`) with jsdocs in typescript...any docs generator would infer the type from the typed parameters below.